### PR TITLE
Share code between Unsafe ByteBuf implementations

### PR DESCRIPTION
**Pull request by normanmaurer**
Motiviation:

We have a lot of duplicated code which makes it hard to maintain.

Modification:

Move shared code to UnsafeByteBufUtil and use it in the implementations.

Result:

Less duplicated code and so easier to maintain.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@nmittler @Scottmitch PTAL

File: buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java [hunk 1]
```diff
@@ -1159,7 +1159,7 @@ protected final void checkDstIndex(int index, int length, int dstIndex, int dstC
         }
     }
 
-    private static boolean isInvalid(int index, int length, int capacity) {
+    static boolean isInvalid(int index, int length, int capacity) {
```
Comment (nmittler) [inline buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java]: Would it make sense to move this to a util class (e.g. ByteBufUtil)?

**Comment by nmittler** [comment]
LGTM ... just one minor comment.

File: buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java (hunk 1 above)
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java]: sure we could... not sure if it's needed thought as I would like to keep it package private :)

Comment (nmittler) [inline buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java]: Yup makes sense.

File: buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java [hunk 2]
```diff
@@ -243,5 +247,146 @@ static void setLong(byte[] array, int index, long  value) {
...
+            }
+            return readBytes;
+        } finally {
+            tmpBuf.release();
+        }
+    }
+
+    static void getBytes(AbstractByteBuf buf, long addr, int index, ByteBuf dst, int dstIndex, int length) {
+        buf.checkIndex(index, length);
+        if (dst == null) {
```
Comment (Scottmitch) [inline buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java]: consider using `checkNotNull`

**Comment by Scottmitch** [comment]
@normanmaurer - 1 small comment (which may apply to a few spots) and lgtm

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (40e0fbfcb6d278fa7a5e92c19debad7dd7b31e08) , 4.1 (2be4bb74a5538b76650b1e6992a8c0eec279f31b) and master (1aba6001b4d9babb3824b9b6fa7bc5feddb9930a)

**Pull request closed without merging**